### PR TITLE
Example: Pachyderm -> Seldon-Core CD4ML

### PR DESCRIPTION
**Pull request by philwinder**
Hi All. This is another pachyderm example where the seldon deployment is build into a pachyderm pipeline.

```release-note
NONE
```

### CODE REVIEW

**Comment by seldondev** [comment]
Hi @philwinder. Thanks for your PR.

I'm waiting for a [SeldonIO](https://github.com/orgs/SeldonIO/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/SeldonIO/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **jklaise**
You can assign the PR to them by writing `/assign @jklaise` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Tue Jan 12 15:42:59 UTC 2021
The logs for [lint] [2] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2833/2.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2833 --build=2

**Comment by seldondev** [comment]
Tue Jan 12 15:43:03 UTC 2021
The logs for [pr-build] [1] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2833/1.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2833 --build=1

**Comment by seldondev** [comment]
@philwinder: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint | 2ba4a8c1849d112818f9868d1452a1b1d8aa8dd0 | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by philwinder** [comment]
/test lint

I believe the previous error was a pip download glitch. Retesting.

**Comment by seldondev** [comment]
Wed Jan 13 10:54:40 UTC 2021
The logs for [lint] [3] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2833/3.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2833 --build=3

File: examples/pachyderm-cd4ml/Makefile
Comment (RafalSkolasinski) [inline examples/pachyderm-cd4ml/Makefile:11]: We would be nice to include `KIND_NAME` that defaults to `kind` as in previous example.

File: examples/pachyderm-cd4ml/index.ipynb
Comment (RafalSkolasinski) [inline examples/pachyderm-cd4ml/index.ipynb:574]: We usually recommend to forward it to port `8003`, as per SC configuration page linked at the top of the notebook:
```
!kubectl port-forward $(kubectl get pods -n seldon-system -l app.kubernetes.io/name=ambassador -o jsonpath='{.items[0].metadata.name}') -n seldon-system 8003:8080
```

(just making a note due to consistency with other exampels as users running it probably already have 8003 port forwarded in separate terminal)

**Review by RafalSkolasinski — CHANGES_REQUESTED** [review]
That's a quite nice example!

I added two minor comments.

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **rafalskolasinski**
You can assign the PR to them by writing `/assign @rafalskolasinski` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Review by RafalSkolasinski — APPROVED** [review]
Nice one!
/approve

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[RafalSkolasinski](https://github.com/SeldonIO/seldon-core/pull/2833#issuecomment-762163368 "Approved")*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com#issuecomment-762163368/SeldonIO/seldon-core/blob/master/OWNERS)~~ [RafalSkolasinski]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
New changes are detected. LGTM label has been removed.

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[RafalSkolasinski](https://github.com/SeldonIO/seldon-core/pull/2833#issuecomment-762163368 "Approved")*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [RafalSkolasinski]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Mon Jan 18 20:40:46 UTC 2021
The logs for [lint] [5] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2833/5.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2833 --build=5

**Comment by seldondev** [comment]
Mon Jan 18 20:40:56 UTC 2021
The logs for [pr-build] [4] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2833/4.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2833 --build=4

**Comment by philwinder** [comment]
Hi @RafalSkolasinski. 

I've just pushed some changes to implement your suggestions and created a dedicated role to create the new SeldonDeployment, rather than using the admin clusterrole which is unsafe.

I have one remaining improvement that I'd like to add, whereby I revert the SeldonDeployment to a previous versioned model, but that's a bit of a bigger job so will take a little more time.

I'm happy to submit that as a separate PR if you want to get this merged.

Thanks, Phil

**Pull request merged**
